### DISCUSSION TITLE
Set Client IP from Session as Environmental Variable

### DISCUSSION THREAD

**Discussion by itxworks** [neutrinolabs/xrdp#2221] [Q&A]
Just wondering if its possible to get the connection ip address into env_set_user 

I tried it like this but no success ...

env_set_user(const char *username, char **passwd_file, int display, char *connection_description,
             const struct list *env_names, const struct list *env_values)  

thx

**Comment by matt335672** [comment]
Excuse the brief reply, as I've got limited time this week.

Yes you can do this, but it will only apply to the session creation. On a reconnect the environment variable will be out-of-date, so not useful - environment variables can't be changed for a whole session once set. For that reason we probably wouldn't put this in the main product.

If you want to do it, the ip address is part of the `connection_description` for the session. You'd need to get this into `env_set_user` either directly, or via the env_names/env_values pairing.

HTH

**Comment by itxworks** [comment]
Hi Matt,

it would be great to get it env_names/env_values -> like g_setenv("REMOTE_CLIENT_IP", connection_description, 1);

as we got it working in an older version ...

sesman/env.c 

```
...
g_clearenv();
g_setenv("SHELL", pw_shell, 1);
g_setenv("PATH", "/bin:/usr/bin:/usr/X11R6/bin:/usr/local/bin", 1);
g_setenv("USER", username, 1);
g_sprintf(text, "%d", uid);
g_setenv("UID", text, 1);
g_setenv("HOME", pw_dir, 1);
g_set_current_dir(pw_dir);
g_sprintf(text, ":%d.0", display);
g_setenv("DISPLAY", text, 1);
g_setenv("REMOTE_CLIENT_IP", client_ip, 1);
if (passwd_file != 0)
...

```

Thank you.

**Comment by matt335672** [comment]
I don't have access to any test facilities this week. Here's an unofficial patch which might do what you want. All I can say is it compiles. I built it on v0.9.19 but it should work on most v0.9.x releases.

```diff
--- a/sesman/session.c
+++ b/sesman/session.c
@@ -587,6 +587,12 @@ session_start_fork(tbus data, tui8 type, struct SCP_SESSION *s)
         }
         else if (window_manager_pid == 0)
         {
+            if (s->connection_description != NULL)
+            {
+                list_add_item(g_cfg->env_names, (tintptr) g_strdup("REMOTE_CLIENT_IP"));
+                list_add_item(g_cfg->env_names, (tintptr) g_strdup(s->connection_description));
+            }
+
             wait_for_xserver(display);
             env_set_user(s->username,
                          0,
```

**Comment by matt335672** [comment]
Oops - obvious problem with the above patch. The second `list_add_item()` call should be for `g_cfg->env_values`, rather than `g_cfg->env_names`.

Here's an improved version that actually works, and just adds the client IP to the environment variable rather than all the other stuff in teh connection description.

```diff
--- a/sesman/session.c
+++ b/sesman/session.c
@@ -587,6 +587,16 @@ session_start_fork(tbus data, tui8 type, struct SCP_SESSION *s)
         }
         else if (window_manager_pid == 0)
         {
+            if (s->connection_description != NULL)
+            {
+                char ip[64];
+                g_get_ip_from_description(s->connection_description,
+                                          ip, sizeof(ip));
+                list_add_item(g_cfg->env_names,
+                             (tintptr) g_strdup("REMOTE_CLIENT_IP"));
+                list_add_item(g_cfg->env_values, (tintptr) g_strdup(ip));
+            }
+
             wait_for_xserver(display);
             env_set_user(s->username,
                          0,
```

I think the correct way to actually meet a requirement for getting the client IP is to add a facility for a session to inquire about the characteristics of the currently connected client. That way you can get the details of the currently client (if any), rather than simply the client which initiated the connection.

This is also related to #392, which is currently unimplemented. It shouldn't be too hard to get that done against devel, actually. I'll have a quick look.

**Comment by itxworks** [comment]
Hi,

great  I got an fail - do I miss something?
```
            if (s->connection_description != NULL) {
                char ip[64];
                g_get_ip_from_description(s->connection_description,
                                          ip, sizeof(ip));
                list_add_item(g_cfg->env_names,
                              (tintptr) g_strdup("REMOTE_CLIENT_IP"));
                list_add_item(g_cfg->env_values, (tintptr) g_strdup(ip));
            }
```
 
```
session.c: In function ‘session_start’:
session.c:594:87: error: expected ‘)’ before ‘;’ token
  594 |                               list_add_item(g_cfg->env_values, (tintptr) g_strdup(ip));
      |                                                                                       ^
      |                                                                                       )
session.c:594:31: error: invalid use of void expression
  594 |                               list_add_item(g_cfg->env_values, (tintptr) g_strdup(ip));
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
session.c:594:88: error: expected ‘;’ before ‘}’ token
  594 |                               list_add_item(g_cfg->env_values, (tintptr) g_strdup(ip));
      |                                                                                        ^
      |                                                                                        ;
  595 |             }
      |             ~
```

**Reply by matt335672** [reply]
I can't actually see anything wrong with that as posted. Check your parentheses in the previous line. Also, try applying the change as a patch rather than typing it in.

**Comment by itxworks** [comment]
You are right I applied the changes as patch an made a fresh install -> build was ok now  ... but sesman triggers an error ...

```
==> /var/log/xrdp-sesman.log <==
[20220412-09:25:59] [ERROR] sesman_main_loop: trans_check_wait_objs failed, removing trans
[20220412-09:26:02] [WARN ] Window manager (pid 115612, display 11) exited quickly (3 secs). This could indicate a window manager config problem
[20220412-09:26:02] [INFO ] Calling auth_stop_session and auth_end from pid 115611
[20220412-09:26:02] [INFO ] Terminating X server (pid 115613) on display 11
[20220412-09:26:02] [INFO ] Terminating the xrdp channel server (pid 115616) on display 11
```
thank you !!!

**Comment by matt335672** [comment]
Suggest you get things working without the parch first, then apply the patch. Preferably, use the one in [this post](https://github.com/neutrinolabs/xrdp/discussions/2221#discussioncomment-2551233).

I can't see anything in the patch itself which should break your window manager. If you're doing a clean build from source, make sure xrdp isn't installed from a package. Check you're not logged in at the console, etc.

**Reply by itxworks** [reply]
You are right .. fresh install from scratch  and now it works as inspected - any plans to add the patch to the regular version?

![image](https://user-images.githubusercontent.com/73403258/163958561-c515f9c3-a591-46d6-8cf5-8cf42ad9338d.png)

thx !!!

**Reply by matt335672** [reply]
There are no plans to add this to the regular version. To summarise my notes above:-

> Yes you can do this, but it will only apply to the session creation. On a reconnect the environment variable will be out-of-date, so not useful - environment variables can't be changed for a whole session once set. For that reason we probably wouldn't put this in the main product.

and

> I think the correct way to actually meet a requirement for getting the client IP is to add a facility for a session to inquire about the characteristics of the currently connected client. That way you can get the details of the currently client (if any), rather than simply the client which initiated the connection.

The second option above may well be added to `xrdp-sesadmin` in the next major version. The move to Unix Domain Sockets we've just implemented in devel will make is possible to simplify the way this command works in the future.

Does that make sense, or am I missing something?

**Comment by HottuDoku** [comment]
Hi,

I've ran into simillar issue. I have Windows 10 clients which connect through RDP to Ubuntu XRDP server. Problem is that these clients need to use barcode scanner. We have out own program that registers this scanner through COM port (rs232) and passes it to information system. This much worked on older settup (Thinlinc client). 

The issue is, that now through XRDP I need to know IP of the client, so I can register the scanner and thats a challenge with xrdp.
I've tried to use patch you suggested without and with slight modifications on my part trying to fix it, but nothing of the sort worked.

The error I've been thrown:
`s->connection_description`
`connection_description` is not a member of `s`

To be clear, Iam running 0.9.19 and my clients are never reconnecting, when they disconned, system kills all of their processes, terminating the session. When they connect again, it creates a new session for them.

Do you have any idea how to resolve this?

**Comment by matt335672** [comment]
Hi @HottuDoku 

I'm a bit confused by your post.

The commit which breaks the above patch is 79bec8110c0e491788f32bf1d9a3faab1ff9d224. This isn't part of v0.9.19 - it's still in the development stream. Are you asking for a patch which works with the development stream, or am I misunderstanding exactly what you are asking?

Thanks.

**Reply by HottuDoku** [reply]
Hi,

sorry for mixing things up a bit. I am actually on `682ca9` and I need exactly [this](https://github.com/neutrinolabs/xrdp/discussions/2221#discussioncomment-2591175) to work. 
I've applied the patch above, but I get error after running `make`:

```
The error I've been thrown from the patched section is:
s->connection_description
connection_description is not a member of s
```
I did try to modify it a little bit, but every time a new error pops out or straight up doesnt do anything.
Is there anyway that I can get around this ?

**Comment by matt335672** [comment]
Here's an updated patch against 682ca95635edd7e388250c612ef7d6b59b7736d5:-

```diff
diff --git a/sesman/session.c b/sesman/session.c
index 1ad30baf..20effc19 100644
--- a/sesman/session.c
+++ b/sesman/session.c
@@ -607,6 +607,13 @@ session_start(long data,
         }
         else if (window_manager_pid == 0)
         {
+            if (s->ip_addr[0] != '\0')
+            {
+                list_add_item(g_cfg->env_names,
+                             (tintptr) g_strdup("REMOTE_CLIENT_IP"));
+                list_add_item(g_cfg->env_values, (tintptr) s->ip_addr);
+            }
+
             wait_for_xserver(display);
             env_set_user(s->username,
                          0,
```

I haven't tested it yet. Please let me know if it does what you want @HottuDoku

**Reply by HottuDoku** [reply]
Works exactly as I needed! Thanks a lot :)

**Reply by metalefty** [reply]
You should keep in mind that the IP address is not necessarily correct as @matt335672 already mentioned. Even if you reconnected to existing xrdp session from different IP address, the environment variable is not updated.

**Reply by itxworks** [reply]
Hello,
I would like to upgrade to latest and it seems like ip_addr[0] is not longer valid  - could you please provide the latest solution thank you ...

**Reply by matt335672** [reply]
Not tested but compiles against 17ba2e01e8faace61ac52cf25b53a80abaa276d5 :-

```diff
--- a/sesman/sesexec/session.c
+++ b/sesman/sesexec/session.c
@@ -226,6 +226,13 @@ start_window_manager(struct login_info *login_info,
 {
     char text[256];
 
+    if (login_info->ip_addr != NULL)
+    {
+        list_add_item(g_cfg->env_names,
+                      (tintptr) g_strdup("REMOTE_CLIENT_IP"));
+        list_add_item(g_cfg->env_values, (tintptr) login_info->ip_addr);
+    }
+
     env_set_user(login_info->uid,
                  0,
                  s->display,
```

**Reply by itxworks** [reply]
Hello, for some reason I  am unable get  the latest development running with Debian 11 (with or without this patch) only on Debian 12 - but I can't risk it by now.

But  0.9.24 would work on my Debian 11  install.
----------
uname -a
Linux nfndeterm 5.10.0-27-amd64 #1 SMP Debian 5.10.205-2 (2023-12-31) x86_64 GNU/Linux

xrdp -V
xrdp 0.9.24
----------

**Reply by matt335672** [reply]
Raise an issue and include useful information so we can look at it.

**Reply by itxworks** [reply]
got it running in the terminal-> xrdp-sesman -n & xrdp -n &  so you patch works exactly as I hoped for 👍 

But no luck over systemd:
```-- Journal begins at Thu 2024-02-01 11:53:34 CET. --
Feb 05 20:41:22 nfndeterm xrdp-sesman[17907]: [INFO ] [main(sesman.c:924)] starting xrdp-sesman with pid 17907
Feb 05 20:41:22 nfndeterm systemd[1]: Starting xrdp daemon...
Feb 05 20:41:22 nfndeterm xrdp-sesman[17907]: [INFO ] [sesman_main_loop(sesman.c:475)] Sesman now listening on /var/run/xrdp/sesman.socket
Feb 05 20:41:22 nfndeterm systemd[1]: Started xrdp daemon.
Feb 05 20:41:22 nfndeterm systemd[1]: xrdp.service: Main process exited, code=killed, status=31/SYS
Feb 05 20:41:22 nfndeterm systemd[1]: xrdp.service: Failed with result 'signal'.
Feb 05 20:41:22 nfndeterm xrdp-sesman[17907]: [INFO ] [sesman_main_loop(sesman.c:524)] sesman_main_loop: sesman asked to terminate
Feb 05 20:41:22 nfndeterm systemd[1]: Stopping xrdp session manager...
Feb 05 20:41:22 nfndeterm systemd[1]: xrdp-sesman.service: Succeeded.
Feb 05 20:41:22 nfndeterm systemd[1]: Stopped xrdp session manager.
```

Thank you.

**Reply by matt335672** [reply]
The 'SYS' is down to xrdp using a system call it's not allowed to. See #2697 

1) What version of xrdp are you using now?
2) What do you get for `sudo grep sig=31 /var/log/syslog`
